llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -3926,6 +3926,92 @@ foldICmpUSubSatOrUAddSatWithConstant(ICmpInst::Predicate Pred,
       ConstantInt::get(Op1->getType(), EquivInt));
 }
 
+static Instruction *
+foldICmpOfCmpIntrinsicWithConstant(ICmpInst::Predicate Pred, IntrinsicInst *I,
+                                   const APInt &C,
+                                   InstCombiner::BuilderTy &Builder) {
+  Intrinsic::ID IID = I->getIntrinsicID();
+  Value *LHS = I->getOperand(0);
+  Value *RHS = I->getOperand(1);
+
+  switch (Pred) {
```

Comment:
This switch seems quite verbose, I feel that these cases can be merged because their bodies are basically identical and only the icmp opcodes differ:
- `ICMP_EQ` with `ICMP_NE`
- `ICMP_SGT` with `ICMP_SLE`
- `ICMP_SGE` with `ICMP_SLT`

Reply:
Also, splitting out the function into either two functions or adding a boolean parameter (or local variable) `IsScmp` would help reduce the noise from all the conditional operators as well.

Reply:
Hopefully it's a bit easier to read now.

Reply:
I'd suggest extracting another function that always returns just the unsigned predicate, and then use `ICmpInst::getSignedPredicate()` to convert it to signed afterwards for scmp.